addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/FadingWiFiLEDDriver.java

```diff
@@ -143,49 +175,132 @@ private void changeState(final InternalLedState newState) throws IOException {
-                            logger.debug("fadeState: {}", fadeState);
+        public void run() {
+            lock.lock();
+            if (fadeStepper <= steps) {
+                currentFadeState = fromState.fade(toState, (double) fadeStepper / steps);
+                lock.unlock();
 
-                            sendLEDData(fadeState, outputStream);
+                logger.debug("currentFadeState: {}", currentFadeState);
+                if (!ledSender.apply(currentFadeState)) {
+                    logger.warn("Failed sending at step {}", fadeStepper);
+                    throw new RuntimeException("Failed sending at step " + fadeStepper);
+                }
+                lastComTime = System.currentTimeMillis();
+            } else {
+                lock.unlock();
+                if (lastComTime < (System.currentTimeMillis() - keepCommOpenForMS)) {
+                    throw new RuntimeException("Reached end step");
+                }
+            }
+            fadeStepper++;
+        }
+    }
 
-                            busySleep(fadeDurationInMs / fadeSteps, lastTime);
+    synchronized private void fadeToState(final InternalLedState newTargetState) throws IOException {
+
+        if (ledUpdateSyncSemaphore.tryAcquire(1)) {
+            final Future<?> future;
+            final Socket socket;
+
+            // Create and Execute a new LED Fader
+            try {
+                socket = new Socket(host, port);
+                socket.setSoTimeout(DEFAULT_SOCKET_TIMEOUT);
+                final DataOutputStream outputStream = new DataOutputStream(socket.getOutputStream());
+                logger.debug("Connected to '{}'", socket);
+
+                // ensure controller is on
+                sendRaw(getBytesForPower(true), outputStream);
+
+                ledfaderThread = new LEDFaderRunner(
+                        currentFaderState,
+                        newTargetState,
+                        fadeSteps,
+                        KEEPCOMMOPENINMS,
+                        (fs) -> {
+                            try {
+                                sendLEDData(fs, outputStream);
+                                currentFaderState = fs;
+                                // logger.info("Current: {} {} {} {}", fs.getR(), fs.getG(), fs.getB(), fs.getWhite());
+                                return true;
+                            } catch (IOException e) {
+                                logger.warn("IOException", e);
+                                return false;
+                            }
+                        });
+                future = faderFuture.scheduleAtFixedRate(ledfaderThread, 0, fadeDurationInMs / fadeSteps, TimeUnit.MILLISECONDS);
+
+                // Wait untill LED Thread has finished, when so cleanup socket and remove ledfaderThread
+                waiterFuture.schedule(() -> {
+                    try {
+                        future.get();
+                    } catch (Exception e) {
```

Comment:
future.get() will simply wait untill LEDFaderRunner decided to stop it's execution.
This can happen for a few reasons for example during an IO Exception or when we reached the end of the fading steps or during an InterruptedException.
I will modify it to 
```
                try {
                    future.get();
                } catch (InterruptedException | ExecutionException e) {
                    // Ignored
                } catch (Exception e) {
                    logger.warn("Exception", e);
                }
```
To make it clear that these are the only two that are expected but when any other exception happens it will be logged. Not sure if this is really the best way?